ykrt/src/location.rs

```diff
@@ -97,88 +29,407 @@ impl State {
+    //              │              │─────────────┐
+    //   reprofile  │   Counting   │             │
+    //  ┌──────────▶│              │◀────────────┘
+    //  │           └──────────────┘    increment
+    //  │             │                 count
     //  │             │ start tracing
     //  │             ▼
-    //  │           ┌────────────────────┐
-    //  │           │   PHASE_TRACING    │
-    //  │           └────────────────────┘
-    //  │             │ check for    ▲
-    //  │             │ stuckness    | still
-    //  │             ▼              | active
-    //  │           ┌────────────────────┐
-    //  │           │                    │  abort   ┌────────────────────┐
-    //  │           │ PHASE_TRACING_LOCK │─────────▶│  PHASE_DONT_TRACE  │
-    //  │           │                    │          └────────────────────┘
-    //  │           └────────────────────┘
+    //  │           ┌──────────────┐
+    //  |           |              | incomplete  ┌─────────────┐
+    //  │           │   Tracing    │────────────▶|  DontTrace  |
+    //  |           |              |             └─────────────┘
+    //  │           └──────────────┘
     //  │             │ start compiling trace
     //  │             │ in thread
     //  │             ▼
-    //  │           ┌────────────────────┐
-    //  │           │                    │
-    //  |           |  PHASE_COMPILING   |
-    //  │           │                    │
-    //  │           └────────────────────┘
-    //  │             │ trace maybe    ▲
-    //  │             │ compiled       | trace not yet
-    //  │             ▼                | compiled
-    //  │           ┌────────────────────┐
-    //  │           │    PHASE_LOCKED    │
-    //  │           └────────────────────┘
-    //  │             │ trace definitely
-    //  │             │ compiled
+    //  │           ┌──────────────┐
+    //  |           |  Compiling   |
+    //  │           └──────────────┘
+    //  │             │
+    //  │             │ trace compiled
     //  │             ▼
-    //  │           ┌────────────────────┐
-    //  │           │                    │──────┐
-    //  │           │   PHASE_COMPILED   │      │
-    //  └───────────│                    │◀─────┘
-    //              └────────────────────┘
+    //  │           ┌──────────────┐
+    //  └───────────│   Compiled   |
+    //              └──────────────┘
+    //
+    // We hope that a Location soon reaches the Compiled state (aka "the happy state") and stays
+    // there.
+    //
+    // The state machine is encoded in a usize in a not-entirely-simple way, as we don't want to
+    // allocate any memory for Locations that do not become hot. The layout is as follows (on a 64
+    // bit machine):
+    //
+    //   bit(s) | 63..3   | 2           | 1         | 0
+    //          | payload | IS_COUNTING | IS_PARKED | IS_LOCKED
+    //
+    // In the Counting state, IS_COUNTING is set to 1, and the parking lot mutex bits are unused
```

Comment:
"and the parking lot mutex bits (IS_PARKED and IS_LOCKED) are unused"

Reply:
Fixed (slightly differently) in 281452f2191374b4c090a5c2eb93c8da7e894b09.